scripts/gen_dockerfile_test.py

```diff
@@ -0,0 +1,220 @@
+
+
+@pytest.mark.parametrize('app_yaml', [
+    # Empty app.yaml
+    '',
+    # Invalid entrypoint
+    'entrypoint: "bad \\n entrypoint"',
+    # Invalid python version
+    'runtime_config:\n python_version: 1',
+    'runtime_config:\n python_version: python2',
+])
+def test_get_app_config_invalid(app_yaml):
+    config_file = 'some_config_file'
+    base_image = 'some_image_name'
+    source_dir = 'some_source_dir'
+    raw_app_config = yaml.safe_load(app_yaml)
+    with pytest.raises(ValueError):
+        gen_dockerfile.get_app_config(
+            raw_app_config, base_image, config_file, source_dir)
+
+
+# Basic AppConfig used below
+_base = gen_dockerfile.AppConfig(
+    base_image='',
+    dockerfile_python_version='',
+    entrypoint='',
+    has_requirements_txt=False
+)
+
+
+@pytest.mark.parametrize('app_config, should_find, test_string', [
+    # Requirements.txt
+    (_base, False, 'ADD requirements.txt'),
+    (_base._replace(has_requirements_txt=True), True,
+     'ADD requirements.txt'),
+    # Entrypoint
+    (_base, False, 'CMD'),
+    (_base._replace(entrypoint='my entrypoint'), True,
+     'CMD my entrypoint'),
+    (_base._replace(entrypoint='exec my entrypoint'), True,
+     'CMD exec my entrypoint'),
+    # Base runtime image
+    (_base._replace(base_image='my_base_runtime_image'), True,
+     'FROM my_base_runtime_image'),
+    # Python version
+    (_base._replace(dockerfile_python_version='_my_version'), True,
+     'python_version=python_my_version'),
+])
+def test_generate_files(app_config, should_find, test_string):
+    result = gen_dockerfile.generate_files(app_config)
+    assert sorted(result.keys()) == sorted(EXPECTED_OUTPUT_FILES)
+    dockerfile = result['Dockerfile']
+    if should_find:
+        assert test_string in dockerfile
+    else:
+        assert test_string not in dockerfile
+
+
+@pytest.mark.parametrize('app', [
+    # Sample from https://github.com/GoogleCloudPlatform/python-docs-samples
+    'hello_world',
+])
+def test_generate_dockerfile_command(tmpdir, testdata_dir, app):
+    """Generates output and compares against a set of golden files.
+
+    Optionally runs 'gcloud app gen-config' and compares against that.
+    """
+    app_dir = os.path.join(testdata_dir, app)
+    temp_dir = os.path.join(str(tmpdir), app)
+    os.mkdir(temp_dir)
+
+    # Copy sample app to writable temp dir, and generate Dockerfile.
+    config_dir = os.path.join(temp_dir, 'config')
+    shutil.copytree(app_dir, config_dir)
+    gen_dockerfile.generate_dockerfile_command(
+        base_image='gcr.io/google-appengine/python',
+        config_file=os.path.join(config_dir, 'app.yaml'),
+        source_dir=config_dir)
+
+    # Compare against golden files
+    golden_dir = os.path.join(testdata_dir, app + '_golden')
+    for filename in EXPECTED_OUTPUT_FILES:
+        compare_file(filename, config_dir, golden_dir)
+
+    # Copy sample app to different writable temp dir, and
```

Comment:
I would suggest just making this a separate test with `@pytest.mark.skipif(not shutil.which('gcloud'), reason='Cloud SDK is not installed')`.